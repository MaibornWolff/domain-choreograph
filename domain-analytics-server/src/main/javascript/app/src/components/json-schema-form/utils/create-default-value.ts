import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: undefined,
  number: undefined,
  integer: undefined,
  boolean: undefined,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): any {
  const type = schema.type || 'any';
  if (type instanceof Array) {
    throw new Error('Array types are not supported');
  } else if (type === 'object') {
    return createDefaultValueForObject(schema);
  } else {
    return defaultValues[type];
  }
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      defaultValue[key] = createDefaultValue(property)
    });
  }
  return defaultValue;
}

