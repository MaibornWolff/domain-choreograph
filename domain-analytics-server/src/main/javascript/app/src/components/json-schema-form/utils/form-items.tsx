import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
