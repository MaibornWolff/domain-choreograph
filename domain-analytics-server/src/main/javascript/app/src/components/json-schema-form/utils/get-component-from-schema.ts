import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
