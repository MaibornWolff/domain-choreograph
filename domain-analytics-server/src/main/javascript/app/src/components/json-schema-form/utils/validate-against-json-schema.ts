import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
