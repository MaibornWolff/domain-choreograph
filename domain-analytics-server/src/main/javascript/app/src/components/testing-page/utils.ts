export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
