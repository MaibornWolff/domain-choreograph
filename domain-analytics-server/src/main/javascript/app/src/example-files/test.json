{
  "nodes": {
    "A": {
      "id": "A",
      "name": "A",
      "scope": "root"
    }
  },
  "dependencies": [],
  "scopes": {
    "root": {
      "id": "root",
      "executionContext": "Application",
      "nodes": [
        "A"
      ]
    }
  },
  "executionContexts": {
    "Application": {
      "id": "Application",
      "scopes": [
        "root"
      ]
    }
  }
}
