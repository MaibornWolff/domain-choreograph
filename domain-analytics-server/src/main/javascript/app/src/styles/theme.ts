import darken from 'polished/lib/color/darken';
import hsl from 'polished/lib/color/hsl';
import lighten from 'polished/lib/color/lighten';
import rgba from 'polished/lib/color/rgba';

const textLight = {
  primary: 'rgba(255, 255, 255, 1)',
  secondary: 'rgba(255, 255, 255, .7)',
  disabled: 'rgba(255, 255, 255, .5)'
};

const textDark = {
  primary: 'rgba(0, 0, 0, .87)',
  secondary: 'rgba(0, 0, 0, .54)',
  disabled: 'rgba(0, 0, 0, .38)',
};

function createColorPalette(hue: number, saturation: number, light: number) {
  const difference = .035;
  return [
    hsl(hue, saturation, light + difference * 4),
    hsl(hue, saturation, light + difference * 3),
    hsl(hue, saturation, light + difference * 2),
    hsl(hue, saturation, light + difference),
    hsl(hue, saturation, light),
    hsl(hue, saturation, light - difference),
    hsl(hue, saturation, light - difference * 2),
    hsl(hue, saturation, light - difference * 3),
    hsl(hue, saturation, light - difference * 4),
  ];
}

export const mainColorIndex = 4;

const primary = createColorPalette(225, .4, .5);

const secondary =  createColorPalette(140, .5, .45);

const animations = {
  duration: {
    toggle: 300
  },
  all: `all 0.3s cubic-bezier(.25,.8,.25,1)`
};

const shadows = {
  0: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  1: '0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)',
};

export const THEME = {
  type: 'light' as 'light' | 'dark',
  shadows,
  colors: {
    primary,
    secondary,
    background: {
      0: '#E0E0E0',
      1: '#F5F5F5',
      2: '#FAFAFA',
      3: '#FFFFFF',
    },
    danger: '#b73a2d',
    text: {
      light: textLight,
      dark: textDark,
      normal: textDark,
      inverse: textLight
    },
    icons: {
      light: textLight,
      dark: textDark,
      normal: textDark,
      inverse: textLight
    },
    divider: {
      normal: rgba(0, 0, 0, .12),
      inverse: rgba(255, 255, 255, .12),
    },
    border: {
      normal: rgba(0, 0, 0, .3),
    },
    graph: {
      edge: lighten(.6, 'black'),
      selectedEdge: lighten(.15, primary[mainColorIndex]),
    },
    selectedNode: primary[mainColorIndex],
  },
  jsonViewTheme: 'rjv-default',
  animations,
  dimensions: {
    borderRadius: {
      normal: '3px',
      big: '5px',
    },
    border: {
      normal: '1px',
      thick: '2px'
    }
  }
};

export type Theme = typeof THEME;

const DARK_THEME: Theme = {
  type: 'dark',
  shadows,
  colors: {
    primary,
    secondary,
    background: {
      0: '#191919',
      1: '#343434',
      2: '#424242',
      3: '#303030',
    },
    text: {
      light: textLight,
      dark: textDark,
      normal: textLight,
      inverse: textDark,
    },
    icons: {
      light: textLight,
      dark: textDark,
      normal: textLight,
      inverse: textDark,
    },
    danger: '#d65246',
    divider: {
      normal: rgba(255, 255, 255, .12),
      inverse: rgba(0, 0, 0, .12),
    },
    border: {
      normal: rgba(255, 255, 255, .3),
    },
    graph: {
      edge: darken(.5, 'white'),
      selectedEdge: primary[mainColorIndex],
    },
    selectedNode: primary[mainColorIndex],
  },
  animations,
  jsonViewTheme: 'bright',
  dimensions: THEME.dimensions
};

export const THEMES = {
  light: THEME,
  dark: DARK_THEME
};
export type ThemeKey = keyof typeof THEMES;
