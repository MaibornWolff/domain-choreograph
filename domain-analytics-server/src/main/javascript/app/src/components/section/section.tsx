import * as React from 'react';
import { ToggleSymbol } from '~components/toggle-symbol/toggle-symbol';
import { Theme } from '~styles/theme';
import { wait } from '~utils/async.utils';
import { styled } from '~utils/styled';

const Head = styled('div')`
  display: flex;
  justify-content: flex-start;
  align-items: center;
  padding: .5rem 0;
  cursor: pointer;
`;

const Header = styled('h2')`
  font-size: 1rem;
  padding: 0;
  margin: 0;
  font-weight: bold;
`;

const Line = styled('div')`
  height: 1px;
  background: ${({ theme }) => theme.colors.text.normal.disabled};
`;

const RightLine = styled(Line)`
  flex: 1 1 0;
  margin-left: .5rem;
`;

const LeftLine = styled(Line)`
  flex: 0 0 1rem;
  margin-right: .5rem;
`;

interface BodyProps {
  isOpen: boolean;
  innerHeight: string;
}

const Body = styled.div<BodyProps>`
  overflow: hidden;
  transition: height ${({ theme }) => theme.animations.duration.toggle}ms ease;
  height: ${({ isOpen, innerHeight }) => isOpen ? innerHeight : '0'};
`;

const ChildrenWrapper = styled('div')`
  box-sizing: border-box;
  padding: .5rem 0 .5rem 1rem;
`;

export interface SectionProps {
  header: string;
  theme: Theme;
  isInitiallyOpen: boolean;
}

export interface SectionState {
  isOpen: boolean;
  bodyHeight: string;
}

export class Section extends React.Component<SectionProps, SectionState> {
  public static defaultProps: Partial<SectionProps> = {
    isInitiallyOpen: true
  };

  public state: SectionState = {
    isOpen: Boolean(this.props.isInitiallyOpen),
    bodyHeight: 'auto',
  };

  private bodyRef: HTMLElement | null | undefined;

  public render() {
    const { header, children } = this.props;
    const { isOpen, bodyHeight } = this.state;

    return (
      <div>
        <Head onClick={this.toggleOpen}>
          <ToggleSymbol isOpen={isOpen}/>
          <LeftLine/>
          <Header>{header}</Header>
          <RightLine/>
        </Head>
        <Body
          isOpen={isOpen}
          innerRef={ref => this.bodyRef = ref}
          innerHeight={bodyHeight}
        ><ChildrenWrapper>{children}</ChildrenWrapper></Body>
      </div>
    );
  }

  private toggleOpen = async () => {
    const { isOpen } = this.state;
    const { theme } = this.props;

    this.setState({ bodyHeight: this.getBodyHeight() });

    const OPEN_DELAY = 1;
    await wait(OPEN_DELAY);
    this.setState({ isOpen: !isOpen });

    await wait(theme.animations.duration.toggle);
    this.setState({ bodyHeight: 'auto' });
  }

  private getBodyHeight(): string {
    if (this.bodyRef) {
      return `${this.bodyRef.scrollHeight}px`;
    }
    return '0';
  }
}
