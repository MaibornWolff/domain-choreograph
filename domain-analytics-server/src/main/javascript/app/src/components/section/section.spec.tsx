import { shallow } from 'enzyme';
import * as React from 'react';
import { THEME } from '~styles/theme';
import { wait } from '~utils/async.utils';
import { renderWithProviders } from '~utils/test.utils';
import { Section } from './section';

describe('Section', () => {
  it('should render', () => {
    expect(renderWithProviders(<Section header={'Hello'} theme={THEME}>Body</Section>)).toMatchSnapshot();
  });

  it('should open on header click', async () => {
    const section = shallow<Section>(
      <Section header={'Hello'} theme={THEME} isInitiallyOpen={false}>Body</Section>,
    );
    const header = section.children().first();
    expect(section.state().isOpen).toBe(false);
    header.props().onClick();
    await wait(10);
    expect(section.state().isOpen).toBe(true);
  });
});
