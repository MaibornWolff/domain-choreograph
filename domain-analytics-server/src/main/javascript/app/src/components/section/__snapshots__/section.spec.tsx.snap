// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section should render 1`] = `
.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: .5rem 0;
  cursor: pointer;
}

.emotion-1 {
  height: 24px;
  width: 24px;
  -webkit-transition: -webkit-transform 300ms ease;
  -webkit-transition: transform 300ms ease;
  transition: transform 300ms ease;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-1 svg {
  fill: rgba(0,0,0,.87);
}

.emotion-3 {
  height: 1px;
  background: rgba(0,0,0,.38);
  -webkit-flex: 0 0 1rem;
  -ms-flex: 0 0 1rem;
  flex: 0 0 1rem;
  margin-right: .5rem;
}

.emotion-5 {
  font-size: 1rem;
  padding: 0;
  margin: 0;
  font-weight: bold;
}

.emotion-7 {
  height: 1px;
  background: rgba(0,0,0,.38);
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  margin-left: .5rem;
}

.emotion-13 {
  overflow: hidden;
  -webkit-transition: height 300ms ease;
  transition: height 300ms ease;
  height: auto;
}

.emotion-11 {
  box-sizing: border-box;
  padding: .5rem 0 .5rem 1rem;
}

<div>
  <div
    class="emotion-9 emotion-10"
  >
    <div
      class="emotion-0 emotion-1 emotion-2"
    >
      [object Object]
    </div>
    <div
      class="emotion-3 emotion-4"
    />
    <h2
      class="emotion-5 emotion-6"
    >
      Hello
    </h2>
    <div
      class="emotion-7 emotion-8"
    />
  </div>
  <div
    class="emotion-13 emotion-14"
  >
    <div
      class="emotion-11 emotion-12"
    >
      Body
    </div>
  </div>
</div>
`;
